jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -331,33 +327,90 @@ object XGBoost extends Serializable {
+    savingRounds.map {
+      savingRound: Int =>
+        val tracker = startTracker(nWorkers, trackerConf)
+        try {
+          val parallelismTracker = new SparkParallelismTracker(sc, timeoutRequestWorkers, nWorkers)
+          val overriddenParams = overrideParamsAccordingToTaskCPUs(params, sc)
+          val boostersAndMetrics = buildDistributedBoosters(partitionedData, overriddenParams,
+            tracker.getWorkerEnvs, savingRound, obj, eval, useExternalMemory, missing, prevBooster)
+          val sparkJobThread = new Thread() {
+            override def run() {
+              // force the job
+              boostersAndMetrics.foreachPartition(() => _)
+            }
+          }
+          sparkJobThread.setUncaughtExceptionHandler(tracker)
+          sparkJobThread.start()
+          val isClsTask = isClassificationTask(params)
+          val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
+          logger.info(s"Rabit returns with exit code $trackerReturnVal")
+          val model = postTrackerReturnProcessing(trackerReturnVal, boostersAndMetrics,
+            sparkJobThread, isClsTask)
+          if (isClsTask){
+            model.asInstanceOf[XGBoostClassificationModel].numOfClasses =
+              params.getOrElse("num_class", "2").toString.toInt
+          }
+          if (boosterTmpPath.nonEmpty && savingRound < round) {
+            prevBooster = model.booster
+            saveTmpBooster(sc, boosterTmpPath, model)
+          }
+          model
+      } finally {
+        tracker.stop()
       }
-      sparkJobThread.setUncaughtExceptionHandler(tracker)
-      sparkJobThread.start()
-      val isClsTask = isClassificationTask(params)
-      val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
-      logger.info(s"Rabit returns with exit code $trackerReturnVal")
-      val model = postTrackerReturnProcessing(trackerReturnVal, boostersAndMetrics,
-        sparkJobThread, isClsTask)
-      if (isClsTask){
-        model.asInstanceOf[XGBoostClassificationModel].numOfClasses =
-          params.getOrElse("num_class", "2").toString.toInt
+    }.last
+  }
+
+  private def extractParamsForCheckpoint(
+      sparkContext: SparkContext,
+      params: Map[String, Any],
+      round: Int): (String, Seq[Int]) = {
+    val boosterTmpPath: String = params.get("booster_tmp_path") match {
+      case None => ""
+      case Some(path: String) => path
+      case _ => throw new IllegalArgumentException("parameter \"booster_tmp_path\" must be" +
+        " an instance of String.")
+    }
+
+    val savingFreq: Int = params.get("saving_frequency") match {
+      case None => 0
+      case Some(freq: Int) => freq
+      case _ => throw new IllegalArgumentException("parameter \"saving_frequency\" must be" +
+        " an instance of Int.")
+    }
+    if (boosterTmpPath.nonEmpty && savingFreq > 0) {
+      val prevRound = getHighestVersion(sparkContext, boosterTmpPath) / 2
+      if (prevRound >= round) {
```

Comment:
delete something in a method named "extractParamsForCheckpoint" seems scary

is this branch just for the case that user want to train x more iterations after they have got a model which is completed after y iterations, where y > x?

if this is the case, did we save a model after successfully training for y iterations? (in https://github.com/dmlc/xgboost/pull/2964/files#diff-095250777d030169ae83818b37ab44b1R361) looks like no

and also, does it mean the user has to name their model as x.model (where x is an integer).....

Reply:
we may want to make this more user-friendly

Reply:
This is actually something else. When user reduces the number of rounds of training, we want to clean up all models with higher version. Please take a look at `TmpBoosterManager.cleanUpHigherVersions`